gitindex/index.go

```diff
@@ -532,6 +539,229 @@ func newIgnoreMatcher(tree *object.Tree) (*ignore.Matcher, error) {
+
+	for _, b := range options.Branches {
+		commit, err := getCommit(repository, options.BranchPrefix, b)
+		if err != nil {
+			return nil, nil, nil, nil, fmt.Errorf("getting last current commit for branch %q: %w", b, err)
+		}
+
+		tree, err := commit.Tree()
+		if err != nil {
+			return nil, nil, nil, nil, fmt.Errorf("getting current git tree for branch %q: %w", b, err)
+		}
+
+		branchToCurrentTree[b] = tree
+	}
+
+	rawURL := options.BuildOptions.RepositoryDescription.URL
+	u, err := url.Parse(rawURL)
+	if err != nil {
+		return nil, nil, nil, nil, fmt.Errorf("parsing repository URL %q: %w", rawURL, err)
+	}
+
+	// TODO: Support repository submodules for delta builds
+	// For this prototype, we are ignoring repository submodules, which means that we can use the same
+	// blob location for all files
+	hackSharedBlobLocation := BlobLocation{
+		Repo: repository,
+		URL:  u,
+	}
+
+	// loop over all branches, calculate the diff between our
+	// last indexed commit and the current commit, and add files mentioned in the diff
+	for _, branch := range existingRepository.Branches {
+		lastIndexedCommit, err := getCommit(repository, "", branch.Version)
+		if err != nil {
+			return nil, nil, nil, nil, fmt.Errorf("getting last indexed commit for branch %q: %w", branch.Name, err)
+		}
+
+		lastIndexedTree, err := lastIndexedCommit.Tree()
+		if err != nil {
+			return nil, nil, nil, nil, fmt.Errorf("getting lasted indexed git tree for branch %q: %w", branch.Name, err)
+		}
+
+		changes, err := lastIndexedTree.Diff(branchToCurrentTree[branch.Name])
+		if err != nil {
+			return nil, nil, nil, nil, fmt.Errorf("generating changeset for branch %q: %w", branch.Name, err)
+		}
+
+		patch, err := changes.Patch()
+		if err != nil {
+			return nil, nil, nil, nil, fmt.Errorf("generating patch from changeset for branch %q: %w", branch.Name, err)
+		}
+
+		for _, fp := range patch.FilePatches() {
+			oldFile, newFile := fp.Files()
+
+			if newFile != nil {
+				// either file is added or renamed, so we need to add the new version to the build
+				file := fileKey{Path: newFile.Path(), ID: newFile.Hash()}
+				repos[file] = hackSharedBlobLocation
+				branchMap[file] = append(branchMap[file], branch.Name)
+			}
+
+			if oldFile == nil {
+				// file added - nothing more to do
+				continue
+			}
+
+			// The file is either modified or deleted. So, we need to add ALL versions
+			// of the old file (across all branches) to the build.
+			for b, currentTree := range branchToCurrentTree {
+				f, err := currentTree.File(oldFile.Path())
+				if err != nil {
+					// the file doesn't exist in this branch
+					if errors.Is(err, object.ErrFileNotFound) {
+						continue
+					}
+
+					return nil, nil, nil, nil, fmt.Errorf("getting hash for file %q in branch %q: %w", oldFile.Path(), b, err)
+				}
+
+				file := fileKey{Path: oldFile.Path(), ID: f.ID()}
+				repos[file] = hackSharedBlobLocation
+				branchMap[file] = append(branchMap[file], b)
```

Comment:
I _think_ there is a bug here.

Let's imagine three versions of a file "foo.txt": foo\_1, foo\_2 and foo\_3. We have two branches main and dev:

|         | main   | dev    |
|------- |------ |------ |
| current | foo\_1 | foo\_2 |
| next    | foo\_2 | foo\_3 |

If I am not mistaken branchMap for foo\_2 will contain `main` and `dev`. But in next foo\_2 doesn't exist in `dev`.

Edit: actually I'm having a hard time understanding why it doesn't just always happen because just above this line you add the old file to repos. I need to play around with the indexer which I'll do today after submitting this review.

Reply:
I believe Stefan's explanation is correct - the ID should be different. 

I added a test to show that the existing behavior should be correct under your scenario: [`11a4e16` (#310)](https://github.com/sourcegraph/zoekt/pull/310/commits/11a4e16c5db4bf6cd9f32f70b64670c4bf77b6cd)